gemma-core/src/main/java/ubic/gemma/core/datastructure/matrix/io/MexMatrixWriter.java

```diff
@@ -0,0 +1,325 @@
+ * <p>
+ * The data is written as a TAR archive containing the following entries for each bioassay: {@code {bioAssayName}/barcodes.tsv},
+ * {@code {bioAssayName}/features.tsv}, {@code {bioAssayName}/matrix.mtx}. If using the directory output, individual
+ * files are compressed and will have a {@code .gz} extension.
+ * @author poirigui
+ */
+@CommonsLog
+@Setter
+public class MexMatrixWriter implements SingleCellExpressionDataMatrixWriter {
+
+    /**
+     * Use Ensembl gene IDs instead of gene symbols.
+     */
+    private boolean useEnsemblIds = false;
+
+    @Override
+    public void write( SingleCellExpressionDataMatrix<?> matrix, Writer stream ) throws IOException {
+        throw new UnsupportedOperationException( "MEX is a binary format as it bundles the files in a TAR archive." );
+    }
+
+    @Override
+    public void write( SingleCellExpressionDataMatrix<?> matrix, OutputStream stream ) throws IOException {
+        write( matrix, null, stream );
+    }
+
+    /**
+     * Write a MEX matrix as a TAR archive to the given output stream.
+     */
+    public void write( SingleCellExpressionDataMatrix<?> matrix, @Nullable Map<CompositeSequence, Set<Gene>> cs2gene, OutputStream stream ) throws IOException {
+        try ( TarArchiveOutputStream aos = new TarArchiveOutputStream( stream ) ) {
+            List<BioAssay> bioAssays = matrix.getSingleCellDimension().getBioAssays();
+            for ( int i = 0; i < bioAssays.size(); i++ ) {
+                BioAssay ba = bioAssays.get( i );
+                try ( ByteArrayOutputStream baos = new ByteArrayOutputStream() ) {
+                    writeBarcodes( matrix.getSingleCellDimension(), i, baos );
+                    TarArchiveEntry entry = new TarArchiveEntry( ba.getName() + "/barcodes.tsv" );
+                    entry.setSize( baos.size() );
+                    aos.putArchiveEntry( entry );
+                    aos.write( baos.toByteArray() );
+                    aos.closeArchiveEntry();
+                }
+                try ( ByteArrayOutputStream baos = new ByteArrayOutputStream() ) {
+                    writeFeatures( matrix, cs2gene, baos );
+                    TarArchiveEntry entry = new TarArchiveEntry( ba.getName() + "/features.tsv" );
+                    entry.setSize( baos.size() );
+                    aos.putArchiveEntry( entry );
+                    aos.write( baos.toByteArray() );
+                    aos.closeArchiveEntry();
+                }
+                try ( ByteArrayOutputStream baos = new ByteArrayOutputStream() ) {
+                    TarArchiveEntry entry = new TarArchiveEntry( ba.getName() + "/matrix.mtx" );
+                    writeMatrix( matrix, i, baos );
+                    entry.setSize( baos.size() );
+                    aos.putArchiveEntry( entry );
+                    aos.write( baos.toByteArray() );
+                    aos.closeArchiveEntry();
+                }
+            }
+        }
+    }
```

Comment:
We also need a variant that writes a matrix to a directory, like the streaming one.